akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```
@@ -0,0 +1,419 @@
+   * outbound direction:
+   *
+   *  - if the local user triggered a re-negotiation, cork all user data until
+   *    that is finished
+   *  - if the outbound direction has been closed, trigger outbound readiness
+   *    based upon HandshakeStatus.NEED_WRAP
+   *
+   * These conditions lead to the introduction of a synthetic TransferState
+   * representing the Engine.
+   */
+
+  var lastHandshakeStatus: HandshakeStatus = _
+
+  val engineNeedsWrap = new TransferState {
+    def isReady = lastHandshakeStatus == NEED_WRAP
+    def isCompleted = false
+  }
+
+  val engineInboundOpen = new TransferState {
+    def isReady = !engine.isInboundDone()
+    def isCompleted = false
+  }
+
+  var corkUser = true
+
+  val userHasData = new TransferState {
+    private val user = inputBunch.inputsOrCompleteAvailableFor(UserIn) || userInChoppingBlock
+    def isReady = !corkUser && user.isReady && lastHandshakeStatus != NEED_UNWRAP
+    def isCompleted = false
+  }
+
+  val transportHasData = inputBunch.inputsOrCompleteAvailableFor(TransportIn) || transportInChoppingBlock
+
+  // bidirectional case
+  val outbound = (userHasData || engineNeedsWrap) && outputBunch.demandAvailableFor(TransportOut)
+  val inbound = transportHasData && outputBunch.demandOrCancelAvailableFor(UserOut)
+
+  // half-closed
+  val outboundHalfClosed = engineNeedsWrap && outputBunch.demandAvailableFor(TransportOut)
+  val inboundHalfClosed = transportHasData && engineInboundOpen
+
+  def completeOrFlush(): Unit =
+    if (engine.isOutboundDone()) nextPhase(completedPhase)
+    else nextPhase(flushingOutbound)
+
+  private def doInbound(isOutboundClosed: Boolean, inboundState: TransferState): Boolean =
+    if (inputBunch.isDepleted(TransportIn) && transportInChoppingBlock.isEmpty) {
+      if (tracing) log.debug("closing inbound")
+      try engine.closeInbound()
+      catch { case ex: SSLException ⇒ outputBunch.enqueue(UserOut, SessionTruncated) }
```

No, the reason why we don't treat it as an error. Netty simply swallows the exception.